feat(common): export function to convert to `Date` object

jonathanlie opened the pull request:
The utility function `toDate` used by `formatDate` function to convert its value param to a `Date` is now available to developers who need it.

Fixes #25480

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?
CommonModule only exports `formatDate` function from `i18n/format_date`. Some developers would love the `toDate` function to be exportable as well.

Issue Number: #25480 

## What is the new behavior?
- CommonModule now also exports `toDate` function from `i18n/format_date`.
- Also added `toDate` function to the list of publicApi docs.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

trotyl commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -641,16 +641,26 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
+ *
+ * @publicApi
  */
 export function toDate(value: string | number | Date): Date {
```
Is `toDate()` really a name for public API? Shouldn't it be `parseDate()` or something for better understanding?

jonathanlie replied:
`parseDate` sounds good. LMK if I need to change the name here

ocombe reviewed (approved):
the PR sounds good to me, but I'll need @kara's review and approval to add this function to our existing API (and also to know if we should change the name)

jonathanlie commented:
Hi there! May I know the status of this PR?

kara commented on packages/common/src/i18n/format_date.ts:
```diff
@@ -641,16 +641,26 @@ function convertTimezoneToLocal(date: Date, timezone: string, reverse: boolean):
+ * @ngModule CommonModule
+ * @description
+ *
  * Converts a value to date.
```
Seems like we could make the return value slightly more explicit:

```suggestion
  * Converts a value to a `Date` object.
```

kara commented:
@jonathanlie Seems like two things need to happen

1) My comment needs to be addressed in the PR
2) @IgorMinar will need to review the public API change

jonathanlie committed:
feat(common): export function to convert to `Date` object

The utility ifunction `toDate` used by `formatDate` function to convert its value param to a `Date` is now available to developers who need it.

Fixes #25480

jonathanlie commented:
@kara Updated the return value description to be more explicit.

jonathanlie commented:
Hi, sorry for bothering! Any updates on this PR?

mhevery committed:
fixup! feat(common): export function to convert to `Date` object

mhevery commented:
@IgorMinar please approve.

IgorMinar commented:
hi @jonathanlie, thanks for the PR, but I'm unfortunately exporting this API is not a good fit for Angular and here is why: the parseDate function is just a general utility function that is not specific to how Angular works, and more importantly, unlike the `formatDate` function, not locale specific - the parsing is supported only for the ISO format or for inputs that are already dates.

We have lots of utility functions like this in Angular and if we were to export them we'd be increasing the API surface of Angular making it less focused on building apps. This would additionally increase the maintenance burden on the team because we would not be free to change the implementation as needed for the framework internals.

In the past we made a few mistakes where we exported a few functions like this and it's not been great for either developers (created a perception of complexity and in some cases increased payload size) or the team (increased maintenance burden).

So thanks again for trying to contribute to Angular but in this case, we will not be accepting this change. Thanks for understanding.

jonathanlie commented:
Understood, we will just extract the function ourselves.
Do close the original issue to prevent confusion for other devs too!
